docs: Cloud Deployment Azure (#15397)
grafana/loki: docs/sources/setup/install/helm/deployment-guides/azure.md

---
title: Deploy the Loki Helm chart on Azure
menuTitle: Deploy on Azure
description: Installing the Loki Helm chart on Azure.
keywords: 
---

# Deploy the Loki Helm chart on Azure

This guide shows how to deploy a minimally viable Loki in **microservice** mode on Azure using the Helm chart. To run through this guide, we expect you to have the necessary tools and permissions to deploy resources on Azure, such as:

- Full access to AKS (Azure Kubernetes Service)
- Full access to Azure Blob Storage
- Sufficient permissions to create federated credentials and roles in Azure AD (Active Directory)

> There are three primary methods for deploying Loki on Azure:
> 1. Hard coding a connection string - this is the simplest method but is not recommended for production environments.
> 2. Manged Identity
> 3. Federated Token

Reviewer: Since these aren't steps, change to bullets.
```suggestion
There are three primary methods for deploying Loki on Azure:

- Hard coding a connection string - this is the simplest method but is not recommended for production environments.
- Manged Identity
- Federated Token
```

{{< /admonition >}}

The minimum requirements for deploying Loki on AKS are: 

- Kubernetes version `1.30` or above.
- `3` nodes for the AKS cluster.
- Instance type depends on your workload. A good starting point for a production cluster in the free tier is `Standard_E2ds_v5` instances and for large production workloads `Standard_D4_v5` instances.

Here is how to create an AKS cluster with the minimum requirements:

```bash
az aks create \
  --resource-group <MY_RESOURCE_GROUP_NAME> \
  --name <MY_AKS_CLUSTER_NAME> \
  --node-count 3 \
  --node-vm-size Standard_E2ds_v5 \
  --generate-ssh-keys \
  --enable-workload-identity \
  --enable-oidc-issuer
```

Note in the above command we have enabled workload identity and OIDC issuer. This is required for the Loki service account to authenticate with Azure AD. If you have already created an AKS cluster, you can enable these features using the following command:

```bash
az aks update \
  --resource-group <MY_RESOURCE_GROUP_NAME> \
  --name <MY_AKS_CLUSTER_NAME> \
  --enable-workload-identity \
  --enable-oidc-issuer
```

> The Azure CLI also allows you to bind the AKS cluster to kubectl. You can do this by running the following command:

Reviewer: ```suggestion
The Azure CLI also lets you bind the AKS cluster to kubectl. You can do this by running the following command:
```

```bash
az aks get-credentials --resource-group <MY_RESOURCE_GROUP_NAME> --name <MY_AKS_CLUSTER_NAME>
```

## Configuring Azure Blob Storage

{{< admonition type="tip" >}}

>  **Consider** using unique bucket names rather than;  `chunk`, `ruler` and `admin`. Although Azure Blog Storage is not directly affected by this [security update](https://grafana.com/blog/2024/06/27/grafana-security-update-grafana-loki-and-unintended-data-write-attempts-to-amazon-s3-buckets/) it is best practise to use unique container names (buckets).

Reviewer: ```suggestion
 Consider using unique bucket names rather than:  `chunk`, `ruler` and `admin`. Although Azure Blog Storage is not directly affected by this [security update](https://grafana.com/blog/2024/06/27/grafana-security-update-grafana-loki-and-unintended-data-write-attempts-to-amazon-s3-buckets/) it is a best practice to use unique container names for buckets.
```

{{< /admonition >}}

> Before deploying Loki, you need to create two Azure storage containers; one to store logs (chunks), the second to store alert rules. You can create the containers using the Azure CLI. Containers must exist inside a storage account: 

Reviewer: ```suggestion
Before deploying Loki, you need to create two Azure storage containers; one to store logs (chunks), the second to store alert rules. You can create the containers using the Azure CLI. Containers must exist inside a storage account.
```

{{< admonition type="note" >}}
GEL customers will require a third container to store the admin data. This container is not required for OSS users.
{{< /admonition >}}

1. Create a storage account:

    ```bash
    az storage account create \
    --name <NAME> \
    --location <REGION> \
    --sku Standard_ZRS \
    --encryption-services blob \
    --resource-group <MY_RESOURCE_GROUP_NAME>
    ```

>     **Replace the placeholders with your desired values.** We assume you have a resource group created.

Reviewer: Since we created a resource group up at line 67, I don't think you need this.
```suggestion
    **Replace the placeholders with your desired values.**
```

The recommended way to authenticate Loki with Azure Blob Storage is to use federated credentials. This method is more secure than hard coding a connection string directly into the Loki configuration. In this next section, we will create an Azure AD role and federated credentials for Loki to allow it to read and write from Azure Blob Storage:

1. Locate the OIDC issuer URL:

    ```bash
    az aks show \
    --resource-group <MY_RESOURCE_GROUP_NAME> \
    --name <MY_AKS_CLUSTER_NAME> \
    --query "oidcIssuerProfile.issuerUrl" \
    -o tsv
    ```
    This command will return the OIDC issuer URL. You will need this URL to create the federated credentials.

1. Generate a `credentials.json` file with the following content:
    ```json
    {
        "name": "LokiFederatedIdentity",
        "issuer": "<OIDC-ISSUER-URL>",
        "subject": "system:serviceaccount:loki:loki",
        "description": "Federated identity for Loki accessing Azure resources",
        "audiences": [
          "api://AzureADTokenExchange"
        ]
    }
    ```
    **Replace `<OIDC-ISSUER-URL>` with the OIDC issuer URL you found in the previous step.** Save this file for later use.

> 1. Next generate a azure directory `app` we will use this to assign our federated credentials to:

Reviewer: ```suggestion
1. Next generate an Azure directory `app`. We will use this to assign our federated credentials to:
```

   ```bash
    az ad app create \
    --display-name loki \
    --query appId \
    -o tsv
   ```
    This will return the app ID. Save this for later use. if you need to find the app ID later you can run the following command:
    ```bash
    az ad app list --display-name loki --query "[].appId" -o tsv
    ```

1. The app requires a service principal to authenticate with Azure AD. Create a service principal for the app:

    ```bash
    az ad sp create --id <APP-ID>
    ```
    **Replace `<APP-ID>` with the app ID you generated in the previous step.**

1. Next assign the federated credentials to the app:

    ```bash
    az ad app federated-credential create \
      --id <APP-ID> \
      --parameters credentials.json 
    ```

>     Replace `<APP-ID>` with the app ID you generated in the previous step. **Make sure you have also updated and saved the `credentials.json` file before running this step.**

Reviewer: **Make sure you have also updated and saved the `credentials.json` file before running this step.**
Maybe make this another step, BEFORE this one?

      res = client.rangeQuery(`{format="${format}"}`, "5m", 1000)
      // Check for successful read
      check(res, { 'successful range query': (res) => res.status == 200 });

      // Wait before next iteration
      sleep(1);
    }

    /**
    * Helper function to get random item from array
    */
    function randomChoice(items) {
      return items[Math.floor(Math.random() * items.length)];
    }
   ```

   **Replace `<EXTERNAL-IP>` with the external IP address of the Loki Gateway service.**

   This script will write logs to Loki and query logs from Loki. It will write logs in a random format between 800KB and 2MB and query logs in a random format over the last 5 minutes.
  
3. Run the test:

    ```bash
    ./k6 run azure-test.js
    ```

    This will run the test and output the results. You should see the test writing logs to Loki and querying logs from Loki.

Now that you have successfully deployed Loki in microservices mode on Microsoft Azure, you may wish to explore the following:

- [Sending data to Loki](https://grafana.com/docs/loki/<LOKI_VERSION/send-data/)
- [Querying Loki](https://grafana.com/docs/loki/<LOKI_VERSION>/query/)

> - [Manage](https://grafana.com/docs/loki/<LOKI_VERSION/operations/)

Reviewer: ```suggestion
- [Manage Loki](https://grafana.com/docs/loki/<LOKI_VERSION/operations/)
```

---
title: Deploy the Loki Helm chart on Azure
menuTitle: Deploy on Azure
description: Installing the Loki Helm chart on Azure.
keywords: 
---

# Deploy the Loki Helm chart on Azure

This guide shows how to deploy a minimally viable Loki in **microservice** mode on Azure using the Helm chart. To run through this guide, we expect you to have the necessary tools and permissions to deploy resources on Azure, such as:

- Full access to AKS (Azure Kubernetes Service)
- Full access to Azure Blob Storage
- Sufficient permissions to create federated credentials and roles in Azure AD (Active Directory)

> There are three primary methods for deploying Loki on Azure:

Author: aaah great catch! Thank you

- Hard coding a connection string - this is the simplest method but is not recommended for production environments.
- Manged Identity
- Federated Token

In this guide, we will use the federated token method to deploy Loki on Azure. This method is more secure than hard coding a connection string and is more suitable for production environments.

## Considerations

{{< admonition type="caution" >}}
This guide was accurate at the time it was last updated on **11th of December, 2024**.  As cloud providers frequently update their services and offerings, as a best practice, you should refer to the [Azure documentation](https://learn.microsoft.com/en-us/azure/?product=popular) before creating your storage account and assigning roles.
{{< /admonition >}}

> - **AD Role:** In this tutorial we will create a role in Azure AD to allow Loki to read and write from Azure Blob Storage. This role will be assigned to the Loki service account. You may want to adjust the permissions based on your requirements.

Reviewer: ```suggestion
- **AD Role:** In this tutorial we will create a role in Azure Active Directory (Azure AD) to allow Loki to read and write from Azure Blob Storage. This role will be assigned to the Loki service account. You may want to adjust the permissions based on your requirements.
```

1. Locate the OIDC issuer URL:

    ```bash
    az aks show \
    --resource-group <MY_RESOURCE_GROUP_NAME> \
    --name <MY_AKS_CLUSTER_NAME> \
    --query "oidcIssuerProfile.issuerUrl" \
    -o tsv
    ```
    This command will return the OIDC issuer URL. You will need this URL to create the federated credentials.

1. Generate a `credentials.json` file with the following content:
    ```json
    {
        "name": "LokiFederatedIdentity",
        "issuer": "<OIDC-ISSUER-URL>",
        "subject": "system:serviceaccount:loki:loki",
        "description": "Federated identity for Loki accessing Azure resources",
        "audiences": [
          "api://AzureADTokenExchange"
        ]
    }
    ```
    Replace `<OIDC-ISSUER-URL>` with the OIDC issuer URL you found in the previous step.

1. Make sure you to save the `credentials.json` file before continuing.

1. Next generate an Azure directory `app`. We will use this to assign our federated credentials to:
   ```bash
    az ad app create \
    --display-name loki \
    --query appId \
    -o tsv
   ```

>     This will return the app ID. Save this for later use. if you need to find the app ID later you can run the following command:

Reviewer: ```suggestion
    This will return the app ID. Save this for later use. If you need to find the app ID later you can run the following command:
```

    ```bash
    az ad app list --display-name loki --query "[].appId" -o tsv
    ```

1. The app requires a service principal to authenticate with Azure AD. Create a service principal for the app:

    ```bash
    az ad sp create --id <APP-ID>
    ```
    Replace `<APP-ID>` with the app ID you generated in the previous step.

1. Next assign the federated credentials to the app:

    ```bash
    az ad app federated-credential create \
      --id <APP-ID> \
      --parameters credentials.json 
    ```
    Replace `<APP-ID>` with the app ID you generated in the previous step.

1. Lastly add a role assignment to the app:

    ```bash
    az role assignment create \
      --role "Storage Blob Data Contributor" \
      --assignee <APP-ID> \
      --scope /subscriptions/<SUBSCRIPTION-ID>/resourceGroups/<RESOURCE-GROUP>/providers/Microsoft.Storage/storageAccounts/<STORAGE-ACCOUNT-NAME>
    ```
    Replace the placeholders with your actual values.

> With the Azure AD role and federated credentials created, you can now proceed to deploying Loki using the Helm chart.

Reviewer: ```suggestion
Now that you have created the Azure AD role and federated credentials, you can proceed to deploying Loki using the Helm chart.
```

## Deploying the Helm chart

The following steps require the use of `helm` and `kubectl`. Make sure you have run the `az` command to bind your AKS cluster to `kubectl`:

```bash
az aks get-credentials --resource-group <MY_RESOURCE_GROUP_NAME> --name <MY_AKS_CLUSTER_NAME>
```

Before we can deploy the Loki Helm chart, we need to add the Grafana chart repository to Helm. This repository contains the Loki Helm chart.

1. Add the Grafana chart repository to Helm:

    ```bash
    helm repo add grafana https://grafana.github.io/helm-charts
    ```
1. Update the chart repository:

    ```bash
    helm repo update
    ```
1. Create a new namespace for Loki:

    ```bash
    kubectl create namespace loki
    ```

> ### Loki Basic Authentication

Reviewer: ```suggestion
### Loki basic authentication
```

 replicas: 0
```

{{< admonition type="caution" >}}
Make sure to replace the placeholders with your actual values.
{{< /admonition >}}

It is critical to define a valid `values.yaml` file for the Loki deployment. To remove the risk of misconfiguration, let's break down the configuration options to keep in mind when deploying to Azure:

- **Loki Config vs. Values Config:**
  - The `values.yaml` file contains a section called `loki`, which contains a direct representation of the Loki configuration file.
  - This section defines the Loki configuration, including the schema, storage, and querier configuration.
  - The key configuration to focus on for chunks is the `storage_config` section, where you define the Azure container name and storage account. This tells Loki where to store the chunks.
  - The `ruler` section defines the configuration for the ruler, including the Azure container name and storage account. This tells Loki where to store the alert and recording rules.
  - For the full Loki configuration, refer to the [Loki Configuration](https://grafana.com/docs/loki/<LOKI_VERSION>/configure/) documentation.

- **Storage:**
  - Defines where the Helm chart stores data.

>   - Set the type to `azure` since we are using Azure Blog Storage.

Reviewer: LOL, this is a great typo.
```suggestion
  - Set the type to `azure` since we are using Azure Blob Storage.
```

  - Configure the container names for the chunks and ruler to match the containers created earlier.
  - The `azure` section specifies the storage account name and also sets `useFederatedToken` to `true`. This tells Loki to use federated credentials for authentication.

- **Service Account:**
  - The `serviceAccount` section is used to define the federated workload identity Loki will use to authenticate with Azure AD.
  - We set the `azure.workload.identity/client-id` annotation to the app ID of the Azure AD app.

- **Gateway:**
  - Defines how the Loki gateway will be exposed.
  - We are using a `LoadBalancer` service type in this configuration.

### Deploy Loki

Now that you have created the `values.yaml` file, you can deploy Loki using the Helm chart.

1. Deploy using the newly created `values.yaml` file:

    ```bash
    helm install --values values.yaml loki grafana/loki -n loki --create-namespace
    ```

>     It is important to create a namespace called `loki` as our federated credentials where generated with the subject value `system:serviceaccount:loki:loki`. This translates to the `loki` service account in the `loki` namespace. This is configurable but make sure to update the federated credentials file first.

Reviewer: ```suggestion
    It is important to create a namespace called `loki` as our federated credentials were generated with the  value `system:serviceaccount:loki:loki`. This translates to the `loki` service account in the `loki` namespace. This is configurable but make sure to update the federated credentials file first.
```

    loki-gateway-5f97f78755-hm6mx           1/1     Running   0          10m
    loki-index-gateway-0                    1/1     Running   0          10m
    loki-index-gateway-1                    1/1     Running   0          10m
    loki-ingester-zone-a-0                  1/1     Running   0          10m
    loki-ingester-zone-b-0                  1/1     Running   0          10m
    loki-ingester-zone-c-0                  1/1     Running   0          10m
    loki-querier-89d4ff448-4vr9b            1/1     Running   0          10m
    loki-querier-89d4ff448-7nvrf            1/1     Running   0          10m
    loki-querier-89d4ff448-q89kh            1/1     Running   0          10m
    loki-query-frontend-678899db5-n5wc4     1/1     Running   0          10m
    loki-query-frontend-678899db5-tf69b     1/1     Running   0          10m
    loki-query-scheduler-7d666bf759-9xqb5   1/1     Running   0          10m
    loki-query-scheduler-7d666bf759-kpb5q   1/1     Running   0          10m
    loki-results-cache-0                    2/2     Running   0          10m
    loki-ruler-0                            1/1     Running   0          10m
    ```
  

> ### Find the Loki Gateway Service

Reviewer: ```suggestion
### Find the Loki gateway service
```

> The Loki Gateway service is a LoadBalancer service that exposes the Loki gateway to the internet. This is where you will write logs to and query logs from. By default NGINX is used as the gateway.

Reviewer: ```suggestion
The Loki gateway service is a load balancer service that exposes the Loki gateway to the internet. This is where you will write logs to and query logs from. By default NGINX is used as the gateway.
```

{{< admonition type="caution" >}}

> The Loki Gateway service is exposed to the internet. We provide basic authentication using a username and password in this tutorial. Refer to the [Authentication](https://grafana.com/docs/loki/<LOKI_VERSION>/operations/authentication/) documentation for more information.

Reviewer: ```suggestion
The Loki gateway service is exposed to the internet. We provide basic authentication using a username and password in this tutorial. Refer to the [Authentication](https://grafana.com/docs/loki/<LOKI_VERSION>/operations/authentication/) documentation for more information.
```

{{< /admonition >}}

> To find the Loki Gateway service, run the following command:

Reviewer: ```suggestion
To find the Loki gateway service, run the following command:
```

```bash
kubectl get svc -n loki
```

> You should see the Loki Gateway service with an external IP address. This is the address you will use to write to and query Loki.

Reviewer: ```suggestion
You should see the Loki gateway service with an external IP address. This is the address you will use to write to and query Loki.
```
